### PR TITLE
Drop attempts to set MIS_TOTAL parameter

### PR DESCRIPTION
**Pull request by peterbarker**
Closes https://github.com/ArduPilot/ardupilot/issues/11413

### CODE REVIEW

**Comment by peterbarker** [comment]
We discussed this at a devcall.  There are more elaborate solutions which would fix the problem more nicely, but this works in the short term.

File: Tools/autotest/common.py [hunk 1]
```diff
@@ -1028,11 +1028,11 @@ def should_fetch_all_for_parameter_change(param_name):
             return True
         return False
 
-    def set_parameter(self, name, value, add_to_context=True, epsilon=0.0002):
+    def set_parameter(self, name, value, add_to_context=True, epsilon=0.0002, retries=10):
         """Set parameters from vehicle."""
         self.progress("Setting %s to %f" % (name, value))
         old_value = self.get_parameter(name, retry=2)
-        for i in range(1, 10):
+        for i in range(1, retries+2):
```
Comment (WickedShell) [inline Tools/autotest/common.py]: Why +2 ? This looks to be trying harder then it was before.

Comment (peterbarker) [reply Tools/autotest/common.py]: This loop now goes all the way to 11.  Need I say more?

**Comment by peterbarker** [comment]
On Mon, 3 Jun 2019, WickedShell wrote:

> I'm not thrilled about the solution, as this is messy from a GCS side as there isn't an ACK and it just looks like packet loss from the protocol.

Yep, I realised this when writing it.

OTOH, if we drop a parameter in ArduPilot you'd get the same effect, so 
GCSs must already cope with this in a not-too-bad fashion.

**Comment by peterbarker** [comment]
This has been modified to notify the GCS that we accepted their set - but instantly send the current value.  This is so we don't break the expected parameter state machine.
```
STABILIZE> param set MIS_TOTAL 17
STABILIZE> APM: Param write denied (MIS_TOTAL)
< PARAM_VALUE {param_id : MIS_TOTAL, param_value : 17.0, param_type : 4, param_count : 1139, param_index : 65535}
< PARAM_VALUE {param_id : MIS_TOTAL, param_value : 3.0, param_type : 4, param_count : 1139, param_index : 65535}

param set RC4_TRIM  1600
STABILIZE> < PARAM_VALUE {param_id : RC4_TRIM, param_value : 1600.0, param_type : 4, param_count : 1139, param_index : 65535}
```

File: Tools/autotest/common.py [hunk 2]
```diff
@@ -2652,6 +2652,18 @@ def run_tests(self, tests):
 
         return True
 
+    def test_parameters(self):
```
Comment (davidbuzz) [inline Tools/autotest/common.py]: test_behaviour_of_misstotal_parameter?

**Comment by tridge** [comment]
better to use a flag bit in param table

File: Tools/autotest/common.py (hunk 2 above)
Comment (peterbarker) [reply Tools/autotest/common.py]: Added a comment and made a subtest instead.  Just wanted somewhere to throw misellaneous tests or our list of tests is going to get ridiculous.

**Comment by peterbarker** [comment]
Reworked to use a new parameter bit indicting a parameter is for internal use only.

**Comment by peterbarker** [comment]
.... so am I going back to the string-matching thing on this one?

File: libraries/AP_Param/AP_Param.h [hunk 3]
```diff
@@ -63,13 +63,18 @@
 // the var_info is a pointer, allowing for dynamic definition of the var_info tree
 #define AP_PARAM_FLAG_INFO_POINTER  (1<<4)
 
+// this parameter is visible to GCS via mavlink but should never be
+// set by anything other than the ArduPilot code responsible for its
+// use.
+#define AP_PARAM_FLAG_INTERNAL_USE_ONLY (1<<6)
```
Comment (OXINARF) [inline libraries/AP_Param/AP_Param.h]: Shouldn't it be `1<<5`?

File: libraries/AP_Mission/AP_Mission.cpp [hunk 4]
```diff
@@ -14,7 +14,7 @@ const AP_Param::GroupInfo AP_Mission::var_info[] = {
     // @Range: 0 32766
     // @Increment: 1
     // @User: Advanced
-    AP_GROUPINFO("TOTAL",  0, AP_Mission, _cmd_total, 0),
+    AP_GROUPINFO_FLAGS("TOTAL",  0, AP_Mission, _cmd_total, 0, AP_PARAM_FLAG_INTERNAL_USE_ONLY),
```
Comment (OXINARF) [inline libraries/AP_Mission/AP_Mission.cpp]: We should add `@ReadOnly: True` to the docs above.

File: libraries/GCS_MAVLink/GCS_Param.cpp [hunk 5]
```diff
@@ -272,12 +272,24 @@ void GCS_MAVLINK::handle_param_set(mavlink_message_t *msg)
...
     }
+
     float old_value = vp->cast_to_float(var_type);
 
+    if (parameter_flags & AP_PARAM_FLAG_INTERNAL_USE_ONLY) {
+        gcs().send_text(MAV_SEVERITY_WARNING, "Param write denied (%s)", key);
+        // echo back the incorrect value so that we fulfull the
+        // parameter state machine requirements:
+        send_parameter_value(key, var_type, packet.param_value);
+        // and then tell everyone what the correct value is:
```
[the remark this answers is not in the document]
Comment (peterbarker) [reply libraries/GCS_MAVLink/GCS_Param.cpp]: Fixed.

File: libraries/AP_Param/AP_Param.h (hunk 3 above)
Comment (peterbarker) [reply libraries/AP_Param/AP_Param.h]: Yes.  No idea what happened there.

File: libraries/AP_Mission/AP_Mission.cpp (hunk 4 above)
Comment (peterbarker) [reply libraries/AP_Mission/AP_Mission.cpp]: Done

**Comment by peterbarker** [comment]
@tridge @WickedShell As discussed/requested on the DevCall, I've fixed this up based on @OXINARF 's review - should be good to go now.

**Pull request merged**